torch/csrc/jit/codegen/cuda/expr_simplifier.cpp

```diff
@@ -0,0 +1,430 @@
+  return type == BinaryOpType::Add || type == BinaryOpType::Mul ||
+      type == BinaryOpType::And || type == BinaryOpType::Or ||
+      type == BinaryOpType::Xor;
+}
+
+// The expression type that represents the flattened ops. For example, if I have
+// out = a + b + 3 + c + 5, then I will have:
+//   FlattenedAssocCommOp {
+//     inputs: [a, b, c]
+//     putputs: [out]
+//     constant term: 8
+//   }
+//
+// TODO: refactor IR printing so that private exprs like this can be printed.
+// This will be very helpful for debugging.
+class FlattenedAssocCommOp : public Expr {
+ public:
+  using Expr::Expr;
+
+  FlattenedAssocCommOp(
+      IrBuilderPasskey passkey,
+      BinaryOpType op,
+      Val* out,
+      std::vector<Val*> terms)
+      : Expr(passkey) {
+    TORCH_CHECK(
+        isAssociativeAndCommutative(op),
+        "Can only flatten associative and commutative ops");
+    addAttribute(
+        IrBuilder::create<Attribute<BinaryOpType>>(passkey.ir_container_, op));
+    addOutput(out);
+    for (auto v : terms) {
+      // Note that `addInput` is overriden in this class.
+      addInput(v);
+    }
+  }
+
+  NVFUSER_DECLARE_CLONE_AND_CREATE
+
+  virtual const char* getOpString() const override {
+    return "FlattenedAssocCommOp";
+  }
+
+  DataType dtype() const {
+    return *output(0)->getDataType();
+  }
+
+  BinaryOpType getOpType() const {
+    return attribute(0)->as<Attribute<BinaryOpType>>()->value;
+  }
+
+  // Constant terms are stored and handled separately from other inputs. For
+  // example, if I have a + b + 3 + c + 5, then the constant term will be 8, and
+  // a, b, c are all inputs.
+  bool hasConstantTerm() const {
+    return attributes().size() == 2;
+  }
+
+  Val* getConstantTerm() const {
+    return attributeVal(1);
+  }
+
+  template <typename T>
+  T getConstantTermValue() const {
+    return *getConstantTerm()->as<Scalar<T>>()->value();
+  }
+
+  // Get a vector of inputs, sorted as the order given by `variables`. Note that
+  // the sorting key is the rightmost variable that an input depends on. For
+  // example, if I have two inputs.
+  // v1 = a * c
+  // v2 = b
+  // and variables is [a, b, c], then v2 < v1 because the rightmost depending
+  // variable of v2 is b, and the rightmost depending variable of v1 is c,
+  // and b < c. So in this example, this function will return [v2, v1].
+  // Tensors are always considered as variables and they are always considered
+  // as the rightmost.
+  std::vector<Val*> sortedInputs(const std::list<ValInfo>& variables) {
+    std::unordered_set<Val*> variables_set;
+    variables_set.reserve(variables.size());
+    for (auto v : variables) {
+      variables_set.emplace(v.variable);
+    }
+    std::vector<Val*> sorted_inputs(inputs().begin(), inputs().end());
+    auto compare = [&](Val* v1, Val* v2) {
+      bool v1_lefter_than_v2 = false;
+      auto deps1 = getSubexprDependency(v1, variables_set);
+      auto deps2 = getSubexprDependency(v2, variables_set);
```

Comment:
Probably not too important of an optimization but seems a little odd to me to recompute these during the sorting.

Reply:
Changed to a precompute + unordered_map lookup